### PR TITLE
Sample-log cache + single per-batch record callback, with a stage-wise rollout evaluation and target-violation share

### LINKED ISSUE #42: Deterministic-equivalent evaluation can overstate policy quality — add a stage-wise rollout evaluation and a target-violation metric
## Summary

Evaluating a trained TS-DDR policy by solving the **full-horizon deterministic
equivalent** with the policy's targets can silently overstate its quality. The coupled
solve re-optimizes all stages jointly and compensates for targets that are not
followable stage by stage — which is exactly what deployment cannot do. A
**stage-wise subproblem rollout** evaluation (the package already has the machinery
via the stage-wise / `simulate_multistage` path) should be the headline metric,
together with a target-violation measure.

## Symptom

Same checkpoint (Bolivia ACP example, 96 stages, LSTM 2×128, constant ×1 auto-base
slack penalty), 40 fixed evaluation scenarios, objective excluding the target-slack
penalty term:

| evaluation | cost | target-violation share |
|---|---|---|
| deterministic equivalent (full horizon) | 306,556 | 0.046 |
| stage-wise rollout (deployment semantics) | **324,437 (+5.8%)** | **0.408** |

The deterministic-equivalent view reports a policy that looks near-optimal and
target-consistent; executed stage by stage, the same policy is +5.8% more expensive
and violates its own targets ~41% of the time.

## Why it matters

Stage-wise execution is the deployment semantics of a target-trajectory policy.
Reporting only deterministic-equivalent numbers can make results look better than the
policy would operate, and hides target-followability failures entirely.

## Proposed fix

1. Provide/document an evaluation helper based on the existing stage-wise rollout
   path, and recommend it as the default reported metric in the examples.
2. Report two numbers per evaluation: (a) operational objective **excluding** the
   target-slack penalty term, and (b) a target-violation share (e.g., the realized
   slack-penalty value normalized by the objective), with the recommendation that
   policy comparisons are only trusted when the violation share is small (≤ ~0.05).
3. Note the interaction with the penalty-annealing issue (#XX): with an annealed
   schedule the two evaluations agree at convergence to ~1e-7 relative — the rollout
   metric is the guard that detects when they don't.

### PR DESCRIPTION
**Pull request by DJLee1208**
Fixes #42

This implements the sample-logger + single per-batch `record` design proposed in the issue thread.

## Summary

Deterministic-equivalent evaluation can overstate policy quality: the coupled solve re-optimizes all stages
jointly and absorbs stage-wise-unfollowable targets through the slack penalty. This PR (1) restructures the
recording API of `train_multistage` around a per-sample sample logger and a single per-batch record call, as
proposed in the issue comment, and (2) ships a ready-made record implementation that evaluates the policy
with a stage-wise rollout (deployment semantics) on a fixed held-out scenario set and reports an
operational-cost metric plus a target-violation share.

## What changed

- `src/utils.jl`: `SampleLog` — a callable struct patterned after `SaveBest` that keeps a per-batch cache;
  the training loops call `sample_log(s, det_equivalent_or_subproblems)` right after each sample's solve.
  By default it caches the full objective and the objective excluding the target-slack penalty; an
  `on_sample(s, models, sample_log)` hook enables sample-wise debugging against the live JuMP models
  (statuses, suspicious-sample dumps) with no default per-sample logging cost.
- `src/simulate_multistage.jl` (both `train_multistage` overloads; positional arguments unchanged): the
  hardcoded `get_objective_no_target_deficit` accumulation inside the gradient closure is replaced by the
  sample-log call, and the two per-batch `record_loss(iter, model, loss, tag)` invocations are replaced by
  one `record(sample_log, iter, model)`; a `true` return stops training, as before.
- Backward compatibility: the default `record` reproduces the historical console output and metric values
  (same floating-point accumulation order). An explicitly passed legacy `record_loss` keyword keeps working
  through a deprecation adapter that preserves the historical two-call short-circuit ("metrics/loss" first;
  a stop there suppresses the "metrics/training_loss" call). Passing both `record` and `record_loss` is an
  error. Untouched examples that pass `record_loss` continue to work.
- `RolloutEvaluation` (`src/utils.jl`): a record implementation that, every `stride` batches, rolls the
  policy out stage by stage over a fixed, materialized held-out scenario set and reports (a) the rollout
  objective excluding the target-slack penalty term (operational cost) and (b) the target-violation share
  (realized slack penalty / full objective; `NaN` when undefined). Comparisons should only be trusted when
  the share is ≤ ~0.05 — documented in the docstring and README.
- `examples/HydroPowerModels/train_dr_hydropowermodels_subproblems.jl` is wired to the new record API
  (checkpointing/stalling still fed the mean full training objective) with the rollout evaluation on a
  fixed held-out set; README gains an "Evaluation" section with a usage snippet.
- `test/runtests.jl`: cache-vs-live-model equality on the package test problem, per-sample hook firing,
  early-stop contract, legacy-adapter call order + short-circuit + both-kwargs error, fixed-set rollout
  metrics with a reachable-target policy (violation share ~0 by construction), and share edge-case guards.
- `train_multiple_shooting` is intentionally not converted in this PR (it has its own out-of-gradient
  evaluation loop); follow-up material if desired.

## Validation

- `Pkg.test()` green on this branch (189/189, includes the new testsets).
- Reduced-scale demo (Bolivia DCP, 12 stages, 100 batches, vanilla Ipopt, fixed 4-scenario held-out set,
  stride 10): during constant-penalty deterministic-equivalent training, `RolloutEvaluation` reports the
  rollout operational cost and the target-violation share live every 10 batches. The det-eq training
  metric falls from 40.6k to 22.8k (with small fluctuations) while the violation share stays far outside
  the trust gate
  (0.95 -> 0.57 over the run) — exactly the situation where the guard tells you not to trust the
  deterministic-equivalent picture yet.

## Notes

- The shape of the recording API changes (single `record(sample_log, iter, model)` per batch instead of two
  4-argument `record_loss` calls) — as proposed by the maintainer in the issue thread; the deprecation
  adapter keeps explicitly-passed `record_loss` callbacks working unchanged.
- Note: this PR and #41/#42 both touch
  `examples/HydroPowerModels/train_dr_hydropowermodels_subproblems.jl`; whichever merges second will be
  rebased onto the other.

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -158,6 +158,186 @@ function (callback::StallingCriterium)(iter, model, loss)
...
+                           scenarios; stride=1, record=default_record)
+    isempty(scenarios) && throw(ArgumentError(
+        "scenarios must be a nonempty vector of materialized scenarios; sample them once before training so every evaluation uses the same fixed set"))
+    stride >= 1 || throw(ArgumentError("stride must be >= 1"))
+    return RolloutEvaluation(subproblems, state_params_in, state_params_out, initial_state,
+        collect(scenarios), stride, record, NaN, NaN)
+end
+
+function (evaluation::RolloutEvaluation)(sample_log, iter, model)
+    stop = evaluation.record(sample_log, iter, model)
```
Comment (andrewrosemberg) [inline src/utils.jl]: Why put sample_log inside `RolloutEvaluation` just to call it with no additional information?
